zulip/zulip/__init__.py

```diff
@@ -628,7 +628,13 @@ def do_register():
         # Make long-polling requests with `get_events`. Once a request
         # has received an answer, pass it to the callback and before
         # making a new long-polling request.
-        while True:
+        # NOTE: Back off exponentially to cover against potential bugs in this
+        #       library causing a DoS attack against a server when getting errors
+        #       (explicit values listed for clarity)
+        backoff = RandomExponentialBackoff(maximum_retries=10,
```

Comment:
Maybe we can go with `maximum_retries=50`, i.e. basically arranging it so that it will eventually give up, but only if it's really clear the server is down for good and not just having temporary downtime.

I also notice there's a `time.sleep(1)` in `do_register`; probably that should be changed as well.

(Or maybe `do_register` should be folded into this loop?)

Reply:
With this PR, the behavior is changed from before, which is a little concerning, though arguably useful - the handler will return after some time, rather than be in an infinite loop, so the caller can decide whether to retry again?

We could put the backoff parameters into the method call, or something like them; that could also enable behavior like previously (regarding the infinite loop), but with backoff?

Reply:
I think it is correct to potentially fail, so the caller knows something is horribly wrong.  We have a bunch of clients that are forever looping and getting 401s during registration, and I don't think that their owners have any sign that anything is going wrong.

I think we should raise an exception if we hit the max retries, since existing code is likely not expecting the function to ever return.  Raising an exception means it won't unexpectedly fall through to other code.

This behavior change certainly merits a documentation update as well.